nebula/ngMeta/MetaClient.py

```diff
@@ -0,0 +1,283 @@
+import random
+import socket
+import struct
+
+from thrift.transport import TTransport
+from thrift.transport import TSocket
+from thrift.protocol import TBinaryProtocol
+
+from meta.MetaService import Client
+from meta.ttypes import EdgeItem
+from meta.ttypes import ErrorCode
+from meta.ttypes import GetEdgeReq
+from meta.ttypes import GetEdgeResp
+from meta.ttypes import GetPartsAllocReq
+from meta.ttypes import GetPartsAllocResp
+from meta.ttypes import GetTagReq
+from meta.ttypes import GetTagResp
+from meta.ttypes import ListHostsReq
+from meta.ttypes import ListHostsResp
+from meta.ttypes import ListEdgesReq
+from meta.ttypes import ListEdgesResp
+from meta.ttypes import ListSpacesReq
+from meta.ttypes import ListSpacesResp
+from meta.ttypes import ListTagsReq
+from meta.ttypes import ListTagsResp
+from meta.ttypes import TagItem
+
+class MetaClient:
+    def __init__(self, addresses, timeout=1000,
+                connectionRetry=3, executionRetry=3):
+        self.addresses = addresses
+        self.timeout = timeout
+        self.connectionRetry = connectionRetry
+        self.executionRetry = executionRetry
+        self.spaceNameMap = {} # map<spaceName, spaceId>
+        self.spacePartLocation = {} # map<spaceName, map<partId, list<address>>>
+        self.spacePartLeader = {} # map<spaceName, map<partId, leader'saddress>>
+        self.spaceTagItems = {} # map<spaceName, map<TagItem.tag_name, TagItem>>
+        self.spaceEdgeItems = {} # map<spaceName, map<edgeItem.edge_name, edgeItem>
+        self.tagNameMap = {} # map<spaceName, map<TagItem.tag_id, TagItem.tag_name>
+        self.edgeNameMap = {}
+        self.metaClient = None
+
+    def connect(self):
+        while self.connectionRetry > 0:
+            code = self.doConnect(self.addresses)
+            if code == 0:
+                return ErrorCode.SUCCEEDED
+            self.connectionRetry -= 1
+        return ErrorCode.E_FAIL_TO_CONNECT
+
+    def doConnect(self, addresses):
+        address = addresses[random.randint(0, len(addresses)-1)]
+        #print('metaClient is connecting to: ', address[0], address[1])
+        host = address[0]
+        port = address[1]
+        tTransport = TSocket.TSocket(host, port)
+        tTransport.setTimeout(self.timeout)
+        tTransport = TTransport.TBufferedTransport(tTransport)
+        tProtocol = TBinaryProtocol.TBinaryProtocol(tTransport)
+        tTransport.open()
+        self.metaClient = Client(tProtocol)
+
+        for spaceIdName in self.listSpaces():
```

Comment:
the schema just get when connect, if the user modifies, the cache not update

Reply:
add a RepeatTimer to update schemas every 2s